judge/views/problem.py

```diff
@@ -201,6 +201,93 @@ def get_context_data(self, **kwargs):
+    template_name = 'problem/vote-ajax.html'
+
+    def get_context_data(self, **kwargs):
+        if not self.object.can_vote(self.request.user):
+            raise Http404()
+
+        context = super().get_context_data(**kwargs)
+
+        try:
+            context['vote'] = ProblemPointsVote.objects.get(voter=self.request.profile, problem=self.object)
+        except ObjectDoesNotExist:
+            context['vote'] = None
+
+        context['max_possible_vote'] = settings.DMOJ_PROBLEM_MAX_USER_POINTS_VOTE
+        context['min_possible_vote'] = max(settings.DMOJ_PROBLEM_MIN_USER_POINTS_VOTE,
+                                           settings.DMOJ_PROBLEM_MIN_PROBLEM_POINTS)
+        return context
+
+    def post(self, request, *args, **kwargs):
+        problem = self.get_object()
+        if not problem.can_vote(request.user):
+            return HttpResponseForbidden(_('Not allowed to vote on this problem.'), content_type='text/plain')
+
+        form = ProblemPointsVoteForm(request.POST)
+        if not form.is_valid():
+            return JsonResponse(form.errors, status=400)
+
+        with transaction.atomic():
+            # Delete any pre-existing votes.
+            ProblemPointsVote.objects.filter(voter=request.profile, problem=problem).delete()
+            vote = form.save(commit=False)
+            vote.voter = request.profile
+            vote.problem = problem
+            vote.save()
+        return JsonResponse({'points': vote.points})
+
+
+class DeleteProblemVote(ProblemMixin, SingleObjectMixin, View):
+    http_method_names = ['options', 'post']  # This disables GET requests, even though ProblemMixin.get exists.
+
+    def post(self, request, *args, **kwargs):
+        problem = self.get_object()
+        if not problem.can_vote(request.user):
+            return HttpResponseForbidden(_('Not allowed to delete votes on this problem.'), content_type='text/plain')
+
+        ProblemPointsVote.objects.filter(voter=request.profile, problem=problem).delete()
+        return HttpResponse('success', content_type='text/plain')
+
+
+class ProblemVoteStats(ProblemMixin, DetailView):
+    context_object_name = 'problem'
+    template_name = 'problem/vote-stats-ajax.html'
+
+    def get_context_data(self, **kwargs):
+        if self.request.in_contest:
+            raise Http404()
+
+        context = super().get_context_data(**kwargs)
+
+        votes = list(self.object.problem_points_votes.order_by('points').values_list('points', flat=True))
+        context['votes'] = votes
+
+        if votes:
+            context['mean'] = sum(votes) / len(votes)
+            if len(votes) % 2 == 1:
+                context['median'] = votes[len(votes) // 2]
+            else:
+                context['median'] = (votes[len(votes) // 2] + votes[len(votes) // 2 - 1]) / 2
+
+        context['max_possible_vote'] = settings.DMOJ_PROBLEM_MAX_USER_POINTS_VOTE
+        context['min_possible_vote'] = max(settings.DMOJ_PROBLEM_MIN_USER_POINTS_VOTE,
```

Comment:
Other people have thoughts on this code duplication? Maybe it doesn't matter, my only concern would be if these calculations were to change.

Reply:
Let's not deal with this.

Reply:
Let's not worry about this. Avoiding duplication is a lot of work.